### PR TITLE
timed: pull time from the modem

### LINKED ISSUE #30702: [$300 bounty] timed: get time and timezone from the modem
The comma 3/3X has a [Quectel EG25-G ](https://www.quectel.com/product/lte-eg25-g), and we should be able to pull the time and timezone from the cell network. If this ends up working well, this will remove the timezoned daemon, a bunch of panda RTC code, and dependence on an RTC battery.

The goal is to write a small Python daemon to get the current time and timezone. Requirements:
* works without a SIM, an inactive SIM, and an active SIM
* must replace the `timezoned` functionality
* must be fast (aside from modem bootup and ModemManager startup time)
* works with ModemManager running
* needs a unit test, [this](https://github.com/commaai/openpilot/blob/master/system/qcomgpsd/tests/test_qcomgpsd.py) is a good example of one already using the modem

### PR DESCRIPTION
**Pull request by dcondrey**
Need assistance testing.

issue #30702

**Description**
Update timezoned daemon to sync from cell network with fallback

### CODE REVIEW

**Review by adeebshihadeh — COMMENTED** [review]
This looks like a good start! Bounty is locked to you.

To make this easier to review, let's leave timezoned alone and just make a new `system/timed.py` file. The parts you need from timezoned can either be imported or copy/pasted. After this is merged, I'll make a new PR that removes timezoned and enables timed.

**Comment by royjr** [comment]
I'm can help test this @dcondrey

**Comment by royjr** [comment]
```
comma@comma-33e0a0c6:/data/openpilot$ mmcli -m any --command AT+CTZR?
response: '+CTZR: 2'
```

[AT+CTZU Automatic Time Zone Update](https://wiki.teltonika-networks.com/images/3/32/Quectel_EC25%26EC21_AT_Commands_Manual_V1.3.pdf#page=83&zoom=100,68,644)
[AT+CTZR Time Zone Reporting](https://wiki.teltonika-networks.com/images/3/32/Quectel_EC25%26EC21_AT_Commands_Manual_V1.3.pdf#page=84&zoom=100,68,745)
[AT+QLTS Obtain the Latest Time SynchronizedThrough Network](https://wiki.teltonika-networks.com/images/3/32/Quectel_EC25%26EC21_AT_Commands_Manual_V1.3.pdf#page=86&zoom=100,68,180)

```
# Query supported network time modes
comma@comma-33e0a0c6:/data/openpilot$ mmcli -m any --command AT+QLTS=?
response: '+QLTS: (0-2)'

# Query the latest time synchronized through network.
comma@comma-33e0a0c6:/data/openpilot$ mmcli -m any --command AT+QLTS=0
response: '+QLTS: "2023/12/14,16:20:45-20,0"'

# Query the current GMT time
comma@comma-33e0a0c6:/data/openpilot$ mmcli -m any --command AT+QLTS=1
response: '+QLTS: "2023/12/14,18:05:48-20,0"'

# Query the current LOCAL time
comma@comma-33e0a0c6:/data/openpilot$ mmcli -m any --command AT+QLTS=2
response: '+QLTS: "2023/12/14,13:05:50-20,0"'
```

**Comment by royjr** [comment]
```
comma@comma-33e0a0c6:/data/openpilot$ mmcli -m 0 --command AT+CTZR?
response: '+CTZR: 0'
```

We can also set timezone with modem, but how do pick from the list of cities `America/New_York` when all we have is the time difference `-5`. See updated working example https://github.com/commaai/openpilot/compare/master...royjr:openpilot:timed

**Comment by adeebshihadeh** [comment]
> We can also set timezone with modem, but how do pick from the list of cities `America/New_York` when all we have is the time difference `-5`. See updated working example [master...royjr:openpilot:timed](https://github.com/commaai/openpilot/compare/master...royjr:openpilot:timed)

I think any city should be fine as long as it's in the right timezone?

**Comment by royjr** [comment]
Seems like it
https://github.com/royjr/openpilot/blob/b5c4829cbef6b4a9c7ab094c1dfa85b51e41b7dd/system/timed.py#L102-L108

I was going to use `"Etc/GMT-2", "Etc/GMT-1", "Etc/GMT", "Etc/GMT+1"..` but it doesn't allow for quarter-hour timezones.
We can use `export TZ='<+0630>-6:30'` to set custom timezones which `date` uses.

```
comma@comma-33e0a0c6:/data/openpilot$ date
Sat 16 Dec 2023 08:45:52 PM EST

comma@comma-33e0a0c6:/data/openpilot$ export TZ='<+0830>-8:30'

comma@comma-33e0a0c6:/data/openpilot$ date
Sun 17 Dec 2023 10:16:33 AM +0830
```

**Comment by royjr** [comment]
Still working on this? @dcondrey

**Comment by adeebshihadeh** [comment]
As stated in the bounty docs, the bounty lock times out when you stop working on it. We'll give this another day before it times out to make it a full week.

**Comment by adeebshihadeh** [comment]
Alright, @royjr this is up for grabs if you want it. Would love to merge this soon

**Comment by dcondrey** [comment]
> Still working on this? @dcondrey

Sorry for the delay, I just updated the code this evening.

File: system/timed.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+def set_time(time_str):
+  """Set system time using a time string in the format 'YYYY/MM/dd hh:mm:ss±zz'."""
+  try:
+    subprocess.run(['date', '-s', time_str], check=True)
+  except subprocess.CalledProcessError as e:
+    cloudlog.error(f"Error setting system time: {e}")
+    
+def get_time_from_modem():
+  """Retrieve modem's time and timezone, returning timezone abbreviation and time string (tuple or None)."""
+  response = send_at_command('AT+CLTS=2')
```
Comment (royjr) [inline system/timed.py:70]:
```
Error sending AT command: Command '['mmcli', '-m', '0', '--command', 'AT+CLTS=2']' returned non-zero exit status 1.
```

**Comment by adeebshihadeh** [comment]
Alright, this one is unlocked again. @royjr want to finish it up?

@dcondrey I'd recommend picking up a bounty that doesn't require special HW; there's plenty that just need a PC: https://github.com/orgs/commaai/projects/26

**Pull request closed without merging**